<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mutation Mapping Pipiline for *C. elegans* EMS mutagenesis and Backcross Experiments • MutantSets</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Mutation Mapping Pipiline for *C. elegans* EMS mutagenesis and Backcross Experiments">
<meta property="og:description" content="MutantSets">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MutantSets</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9002</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Generating-input.html">Mutation Mapping Pipiline for *C. elegans* EMS mutagenesis and Backcross Experiments</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Mutation Mapping Pipiline for <em>C. elegans</em> EMS mutagenesis and Backcross Experiments</h1>
                        <h4 data-toc-skip class="author">Richard J. Acton</h4>
            
            <h4 data-toc-skip class="date">2022-07-11</h4>
      
      
      <div class="hidden name"><code>Generating-input.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<ul>
<li>
<strong>Inputs:</strong>
<ul>
<li>paired-end fastq files to a galaxy <span class="citation">(<a href="#ref-Afgan2016" role="doc-biblioref">Afgan et al. 2016</a>; <a href="#ref-Jalili2020" role="doc-biblioref">Jalili et al. 2020</a>)</span> history as <code>list of dataset pairs</code>
</li>
<li>A suitable genome fasta file (<em>C. elegans</em>, ce11.fa.gz - Compatible with WBcel235.86 used by SnpEff)</li>
</ul>
</li>
<li>Run the Pipeline: <a href="https://usegalaxy.eu/u/richardjacton/w/c-elegans-ems-mutagenesis-mutation-caller" class="external-link uri">https://usegalaxy.eu/u/richardjacton/w/c-elegans-ems-mutagenesis-mutation-caller</a>
</li>
<li>
<strong>Outputs:</strong>
<ul>
<li>
<a href="https://multiqc.info/" class="external-link"><code>MultiQC</code></a> HTML report with QC info on the input fastqs, trimming, mapping, and deduplication steps.</li>
<li>
<code>.vcf</code> file with variants from all samples (FreeBayes mutation caller)</li>
<li>
<code>.vcf</code> file with variants from all samples (MiModD mutation caller)</li>
<li>
<code>.gff</code> file with deletions from all samples (MiModD deletion calling tool)</li>
</ul>
</li>
<li>Perform Quality filtering and appropriate set subtractions with <a href="https://github.com/RichardJActon/MutantSets" class="external-link"><code>MutantSets</code></a> or alternatively the <code>MiModD VCF Filter</code> or <code>SnpSift Filter</code> tools to identify candidate variants.</li>
<li>(optionally) <code>MiModD NacreousMap</code> for visualisation of mutation locations and <code>MiModD Report Variants</code> for HTML mutation list</li>
</ul>
<p>NB samples are expected to be of the form ‘A123_0001_S1_R1_L001.fq.gz’, sample Identifiers are extracted from this with a regular expression: <code>\w+_(\d+)_S\d+_L\d+.*</code>. This would yield the sample identifier of: 0001. If your file does not conform to this pattern you may need to update this regex by editing the rules in the ‘apply rule to collection’ step of the workflow.</p>
</div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>Doitsidou et al. reviewed Sequencing-Based Approaches for Mutation Mapping and Identification in <em>C. elegans</em> <span class="citation">(<a href="#ref-Doitsidou2016" role="doc-biblioref">Doitsidou, Jarriault, and Poole 2016</a>)</span>. They describe three main approaches to mapping by sequencing:</p>
<ol style="list-style-type: decimal">
<li>Hawaiian variant mapping</li>
<li>EMS-density mapping</li>
<li>Variant discovery mapping</li>
</ol>
<p>This pipeline is currently only compatible with 2 of them, EMS-density mapping &amp; Variant discovery mapping (VDM).</p>
<p><img src="graphics/Doitsidou2016_fig2.png"></p>
<p><img src="graphics/Doitsidou2016_fig3.png"></p>
<div class="section level3">
<h3 id="research-need">Research Need<a class="anchor" aria-label="anchor" href="#research-need"></a>
</h3>
<p>The Schumacher lab identified a need for an analysis pipeline to map and identify mutations in Ethyl methanesulfonate (EMS) mutagenesis forward genetic screens.</p>
<p>Previously a tool called <code>CloudMap</code> <span class="citation">(<a href="#ref-Minevich2012" role="doc-biblioref">Minevich et al. 2012</a>)</span> had been used for this purpose on a Galaxy server. <code>CloundMap</code> is no longer under active development and has been deprecated from <a href="https://usegalaxy.eu/" class="external-link">Galaxy Europe</a> and replaced by <code>MiModD</code> <a href="https://mimodd.readthedocs.io/en/doc0.1.8/index.html" class="external-link">Docs</a></p>
</div>
<div class="section level3">
<h3 id="choice-of-tools">Choice of Tools<a class="anchor" aria-label="anchor" href="#choice-of-tools"></a>
</h3>
<p>In a comparison of <em>C. elegans</em> mutation calling pipelines Smith et al. <span class="citation">(<a href="#ref-Smith2017a" role="doc-biblioref">Smith and Yun 2017</a>)</span> indicated that they had good results with the <code>FreeBayes</code> <span class="citation">(<a href="#ref-Garrison2012" role="doc-biblioref">Garrison and Marth 2012</a>)</span>. So I have initially included this tool here in addition to the<code>MiModD</code> mutation caller to evaluate their relative performance. They also found the the <code>BBMap</code> aligner yielded better results however this is not available in Galaxy so I have opted for <code>Bowtie2</code> for expediency.</p>
</div>
</div>
<div class="section level2">
<h2 id="pipeline-summary">Pipeline Summary<a class="anchor" aria-label="anchor" href="#pipeline-summary"></a>
</h2>
<p><a href="assets/Galaxy-Workflow-EMS_Mutagenesis_Backcross_Mutation_Caller.ga">Pipeline File (Local)</a></p>
<p><a href="https://usegalaxy.eu/u/richardjacton/w/c-elegans-ems-mutagenesis-mutation-caller" class="external-link uri">https://usegalaxy.eu/u/richardjacton/w/c-elegans-ems-mutagenesis-mutation-caller</a></p>
<ul>
<li>Adapter and Quality Trimming with <a href="https://github.com/OpenGene/fastp" class="external-link"><code>fastp</code></a> <span class="citation">(<a href="#ref-Chen2018" role="doc-biblioref">Chen et al. 2018</a>)</span>
</li>
<li>Alignment with <a href="https://github.com/BenLangmead/bowtie2" class="external-link"><code>bowtie2 --sensitive-local</code></a> <span class="citation">(<a href="#ref-Langmead2012" role="doc-biblioref">Langmead and Salzberg 2012</a>)</span>
</li>
<li>
<a href="https://github.com/samtools/samtools" class="external-link"><code>samtools view</code></a> requiring that reads are mapped in a proper pair <span class="citation">(<a href="#ref-Li2009b" role="doc-biblioref">Li et al. 2009</a>)</span>
</li>
<li>Removal of PCR duplicates with <a href="https://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates" class="external-link"><code>Picard MarkDuplicates</code></a> <span class="citation">(<a href="#ref-BroadInstitute2019" role="doc-biblioref"><span>“Picard Toolkit”</span> 2019</a>)</span>
</li>
<li>Left alignment of indels in the BAM files using <a href="https://github.com/ekg/freebayes" class="external-link"><code>FreeBayes</code></a> <span class="citation">(<a href="#ref-Garrison2012" role="doc-biblioref">Garrison and Marth 2012</a>)</span>
</li>
<li>
<a href="https://github.com/ewels/MultiQC" class="external-link"><code>MultiQC</code></a> aggregating quality metrics from trimming, deduplication and alignment <span class="citation">(<a href="#ref-Ewels2016" role="doc-biblioref">Ewels et al. 2016</a>)</span>
</li>
<li>Variant calling with <a href="https://github.com/ekg/freebayes" class="external-link"><code>FreeBayes</code></a> <span class="citation">(<a href="#ref-Garrison2012" role="doc-biblioref">Garrison and Marth 2012</a>)</span>, <a href="https://mimodd.readthedocs.io/" class="external-link"><code>MiModD</code></a> <span class="citation">(<a href="#ref-Baumeister2013" role="doc-biblioref"><span>“MiModD”</span> 2013</a>)</span> variant caller and deletion caller</li>
<li>SNP effect annotation with <a href="http://snpeff.sourceforge.net/SnpEff.html" class="external-link"><code>SnfEff eff</code></a> <span class="citation">(<a href="#ref-Cingolani2012" role="doc-biblioref">Cingolani et al. 2012</a>)</span>
</li>
<li>SNP type annotation with <a href="http://snpeff.sourceforge.net/SnpSift.html" class="external-link"><code>SnpSift Variant Type</code></a> <span class="citation">(<a href="#ref-Cingolani2012" role="doc-biblioref">Cingolani et al. 2012</a>)</span>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="instructions-step-by-step">Instructions (Step-by-Step)<a class="anchor" aria-label="anchor" href="#instructions-step-by-step"></a>
</h2>
<p><strong>1. Upload Data to galaxy</strong></p>
<p><img src="graphics/upload.png"></p>
<p><strong>2. Select all fastq files and create a paired list</strong></p>
<p><img src="graphics/select-all.png"></p>
<p><img src="graphics/build-list-paired-data.png"></p>
<p><strong>3. Pair the fastq files</strong></p>
<p><img src="graphics/pairing-dialog.png"></p>
<p><strong>4. Import the <a href="https://usegalaxy.eu/u/richardjacton/w/c-elegans-ems-mutagenesis-mutation-caller" class="external-link">workflow</a></strong></p>
<p><img src="graphics/import_workflow.png"></p>
<p><strong>5. Run the workflow</strong></p>
<p>Select the paired list object and a genome sequence file as inputs</p>
<p><img src="graphics/run_workflow.png"></p>
<p><strong>6. Check Quality Control Information</strong></p>
<p>Inspect the <code>MultiQC</code> output for signs of technical problems with your data. Consult with your friendly local bioinformatician if there are QC issues you can’t diagnose.</p>
<p><img src="graphics/view_multiqc.png"></p>
<p><strong>7. Preliminary quality filtering <code>SnpSift filter</code></strong></p>
<p>Locate the <a href="https://pcingola.github.io/SnpEff/SnpSift.html#filter" class="external-link"><code>SnpSift filter</code></a> tool in the galaxy tools panel and apply some initial quality filters, simply <code>( QUAL &gt; 15)</code> or <code>20</code> is probably sufficient. Starting with a low stringency filter and applying more stringent criteria when inspecting your candidate mutations it is probably advisable to avoid throwing out possible mutations. Some initial filtering is advisable as the full-sized VCF files may be too large to be easily read by the candidate mutant inspection tool in the next steps. You can check how many lines are in your VCF files by selecting them in the Galaxy history.</p>
<p><img src="graphics/SnpSift_filter.png"></p>
<p><strong>8. Download Data</strong></p>
<p>The main <code>FreeBayes</code> VCF file:</p>
<p><img src="graphics/Download_filtered_vcf.png"></p>
<p>The <code>MiModD</code> deletion calls:</p>
<p><img src="graphics/Download_MiModD_deletions.png"></p>
<p><strong>9. Load the results in the <code>MutantSets</code> Shiny App to identify candidate mutations</strong></p>
<p>If running the App locally, install the <a href="https://www.r-project.org/" class="external-link"><code>R</code></a> package from: <a href="https://github.com/RichardJActon/MutantSets" class="external-link uri">https://github.com/RichardJActon/MutantSets</a></p>
<p>R package installation and running the app locally:</p>
<pre><code><span><span class="co"># install.packages("remotes") # If you don't already have remotes/devtools</span></span>
<span><span class="co"># remotes::install_github("knausb/vcfR") # If vcfR fails to install from CRAN</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"RichardJActon/MutantSets"</span><span class="op">)</span></span>
<span><span class="fu">MutantSets</span><span class="fu">::</span><span class="fu"><a href="../reference/launchApp.html">launchApp</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># opens the app in a web browser</span></span></code></pre>
<ul>
<li>Load the VCF and (optionally) the gff deletion mutant files into <code>MutantSets</code>
</li>
<li>(Optionally) Name your samples something easier to understand</li>
<li>Use the genotype filters to subtract the appropriate sets</li>
<li>Tweak quality and allele frequency thresholds to get a small set of high quality candidates</li>
<li>Assess the candidate mutations by clicking on them and looking at their predicted effects and genomic locations</li>
<li>Download your top results as a <code>.tsv</code> file (openable in excel)</li>
</ul>
<p><strong>You should now have some candidate mutants to screen - Good Luck!</strong></p>
</div>
<div class="section level2">
<h2 id="feedback">Feedback<a class="anchor" aria-label="anchor" href="#feedback"></a>
</h2>
<p>Please direct bug reports, feature requests, and questions to the maintainer of the mutant sets package via [github issues](<a href="https://github.com/RichardJActon/MutantSets/issues" class="external-link uri">https://github.com/RichardJActon/MutantSets/issues</a>.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Afgan2016" class="csl-entry">
Afgan, Enis, Dannon Baker, Marius van den Beek, Daniel Blankenberg, Dave Bouvier, Martin Čech, John Chilton, et al. 2016. <span>“The Galaxy Platform for Accessible, Reproducible and Collaborative Biomedical Analyses: 2016 Update.”</span> <em>Nucleic Acids Research</em> 44 (W1): W3–10. <a href="https://doi.org/10.1093/nar/gkw343" class="external-link">https://doi.org/10.1093/nar/gkw343</a>.
</div>
<div id="ref-Chen2018" class="csl-entry">
Chen, Shifu, Yanqing Zhou, Yaru Chen, and Jia Gu. 2018. <span>“Fastp: An Ultra-Fast All-in-One FASTQ Preprocessor.”</span> <em>Bioinformatics</em> 34 (17): i884–90. <a href="https://doi.org/10.1093/bioinformatics/bty560" class="external-link">https://doi.org/10.1093/bioinformatics/bty560</a>.
</div>
<div id="ref-Cingolani2012" class="csl-entry">
Cingolani, Pablo, Adrian Platts, Le Lily Wang, Melissa Coon, Tung Nguyen, Luan Wang, Susan J. Land, Xiangyi Lu, and Douglas M. Ruden. 2012. <span>“A Program for Annotating and Predicting the Effects of Single Nucleotide Polymorphisms, SnpEff.”</span> <em>Fly</em> 6 (2): 80–92. <a href="https://doi.org/10.4161/fly.19695" class="external-link">https://doi.org/10.4161/fly.19695</a>.
</div>
<div id="ref-Doitsidou2016" class="csl-entry">
Doitsidou, Maria, Sophie Jarriault, and Richard J. Poole. 2016. <span>“Next-Generation Sequencing-Based Approaches for Mutation Mapping and Identification in Caenorhabditis Elegans.”</span> <em>Genetics</em> 204 (2): 451–74. <a href="https://doi.org/10.1534/genetics.115.186197" class="external-link">https://doi.org/10.1534/genetics.115.186197</a>.
</div>
<div id="ref-Ewels2016" class="csl-entry">
Ewels, Philip, Måns Magnusson, Sverker Lundin, and Max Käller. 2016. <span>“MultiQC: Summarize Analysis Results for Multiple Tools and Samples in a Single Report.”</span> <em>Bioinformatics (Oxford, England)</em> 32 (19): 3047–48. <a href="https://doi.org/10.1093/bioinformatics/btw354" class="external-link">https://doi.org/10.1093/bioinformatics/btw354</a>.
</div>
<div id="ref-Garrison2012" class="csl-entry">
Garrison, Erik, and Gabor Marth. 2012. <span>“Haplotype-Based Variant Detection from Short-Read Sequencing,”</span> July, 1–9. <a href="http://arxiv.org/abs/1207.3907" class="external-link">http://arxiv.org/abs/1207.3907</a>.
</div>
<div id="ref-Jalili2020" class="csl-entry">
Jalili, Vahid, Enis Afgan, Qiang Gu, Dave Clements, Daniel Blankenberg, Jeremy Goecks, James Taylor, and Anton Nekrutenko. 2020. <span>“The Galaxy Platform for Accessible, Reproducible and Collaborative Biomedical Analyses: 2020 Update.”</span> <em>Nucleic Acids Research</em> 48 (W1): W395–402. <a href="https://doi.org/10.1093/nar/gkaa434" class="external-link">https://doi.org/10.1093/nar/gkaa434</a>.
</div>
<div id="ref-Langmead2012" class="csl-entry">
Langmead, Ben, and Steven L Salzberg. 2012. <span>“Fast Gapped-Read Alignment with Bowtie 2.”</span> <em>Nature Methods</em> 9 (4): 357–59. <a href="https://doi.org/10.1038/nmeth.1923" class="external-link">https://doi.org/10.1038/nmeth.1923</a>.
</div>
<div id="ref-Li2009b" class="csl-entry">
Li, H., B. Handsaker, A. Wysoker, T. Fennell, J. Ruan, N. Homer, G. Marth, G. Abecasis, and R. Durbin. 2009. <span>“The Sequence Alignment/Map Format and SAMtools.”</span> <em>Bioinformatics</em> 25 (16): 2078–79. <a href="https://doi.org/10.1093/bioinformatics/btp352" class="external-link">https://doi.org/10.1093/bioinformatics/btp352</a>.
</div>
<div id="ref-Baumeister2013" class="csl-entry">
<span>“MiModD.”</span> 2013. Baumeister Lab. <a href="https://sourceforge.net/projects/mimodd/" class="external-link">https://sourceforge.net/projects/mimodd/</a>.
</div>
<div id="ref-Minevich2012" class="csl-entry">
Minevich, Gregory, Danny S. Park, Daniel Blankenberg, Richard J. Poole, and Oliver Hobert. 2012. <span>“CloudMap: A Cloud-Based Pipeline for Analysis of Mutant Genome Sequences.”</span> <em>Genetics</em> 192 (4): 1249–69. <a href="https://doi.org/10.1534/genetics.112.144204" class="external-link">https://doi.org/10.1534/genetics.112.144204</a>.
</div>
<div id="ref-BroadInstitute2019" class="csl-entry">
<span>“Picard Toolkit.”</span> 2019. Broad Institute. <a href="http://broadinstitute.github.io/picard/" class="external-link">http://broadinstitute.github.io/picard/</a>.
</div>
<div id="ref-Smith2017a" class="csl-entry">
Smith, Harold E., and Sijung Yun. 2017. <span>“Evaluating Alignment and Variant-Calling Software for Mutation Identification in c. Elegans by Whole-Genome Sequencing.”</span> Edited by David J. Reiner. <em>PLOS ONE</em> 12 (3): e0174446. <a href="https://doi.org/10.1371/journal.pone.0174446" class="external-link">https://doi.org/10.1371/journal.pone.0174446</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Richard J. Acton.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
